### PR TITLE
Fix NaN gradients in index_reduce/scatter_reduce backward for amax/amin

### LINKED ISSUE #168358: index_reduce returns NaN during backward when compiled and using 'amax'/'amin'
### Description:
When using `torch.Tensor.index_reduce` with the "amax" (or "amin") reduction, eager mode computes correct gradients, but the `torch.compile` version returns incorrect / NaN gradients.
Forward results match, but backward results do not.

### Expected Behavior:
- Forward outputs from eager and compiled should match.
- Gradients should not contain NaNs.

### Actual Behavior
- Forward outputs match.
- Backward gradients diverge:
   - Eager mode produces correct gradients
   - Compiled mode returns NaNs (or otherwise incorrect values)

### To Reproduce 

```python
import torch
torch._inductor.config.fallback_random = True
torch._dynamo.utils.same


def reduce_max(src, index, dim_size, dim=0):
    out_shape = list(src.shape)
    out_shape[dim] = dim_size
    out = torch.full(
        out_shape,
        float("-inf"),
        dtype=src.dtype,
        device=src.device,
    )

    return out.index_reduce(
        dim,
        index,
        src,
        "amax",
        include_self=False,
    )


src = torch.randn(5).to("cuda")
index = torch.zeros(5, dtype=torch.int64).to("cuda")
dim_size = index.max() + 1

# eager test
src.requires_grad_(True)
sm1 = reduce_max(src, index, dim_size, dim=0)
dsm_dr1 = torch.autograd.grad(
    sm1.sum(),
    src,
)[0]
src = src.detach()

# compiled test
comp_scatter_max = torch.compile(reduce_max)
src.requires_grad_(True)
sm2 = comp_scatter_max(src, index, dim_size, dim=0)
dsm_dr2 = torch.autograd.grad(
    sm2.sum(),
    src,
)[0]
src = src.detach()

print("src: ", src)
print("-"*40)
print(f"Results match: {torch.allclose(sm1, sm2)}")
print("Non compiled: ",sm1)
print("Compiled: ", sm2)
print("-"*40)
print(f"Gradients match: {torch.allclose(dsm_dr1, dsm_dr2)}")
print("Non compiled: ", dsm_dr1)
print("Compiled: ", dsm_dr2)
print("-"*40)
```

### Workaround
Defining a custom autograd function with an explicit backward implementation 
produces correct gradients (matching eager mode), even when compiled.

```python
class SafeReduceMax(torch.autograd.Function):
    @staticmethod
    def forward(ctx, src, index, dim_size, dim=0):
        out_shape = list(src.shape)
        out_shape[dim] = dim_size
        out = torch.full(
            out_shape,
            float("-inf"),
            dtype=src.dtype,
            device=src.device,
        )
        out = out.index_reduce(dim, index, src, "amax", include_self=False)
        ctx.save_for_backward(src, index, out)
        ctx.dim = dim
        return out

    @staticmethod
    def backward(ctx, grad_output):
        src, index, out = ctx.saved_tensors
        dim = ctx.dim
        # Identify which src positions contributed to the max
        mask = (src == out.index_select(dim, index)).float()
        # If multiple maxima exist, split gradients evenly
        count = torch.zeros_like(out).index_add(dim, index, mask)
        count = count.index_select(dim, index).clamp(min=1)
        grad_src = mask * (grad_output.index_select(dim, index) / count)

        return grad_src, None, None, None


def safe_reduce_max(src, index, dim_size, dim=0):
    return SafeReduceMax.apply(src, index, dim_size, dim)
```

### Error logs

```bash
MYPATH/:17: UserWarning: index_reduce() is in beta and the API may change at any time. (Triggered internally at /pytorch/aten/src/ATen/native/cuda/Indexing.cu:1432.)
  return out.index_reduce(
src:  tensor([-0.9539, -0.5972, -0.0143, -2.0441, -0.2046], device='cuda:0')
----------------------------------------
Results match: True
Non compiled:  tensor([-0.0143], device='cuda:0', grad_fn=<IndexReduceBackward0>)
Compiled:  tensor([-0.0143], device='cuda:0', grad_fn=<CompiledFunctionBackward>)
----------------------------------------
Gradients match: False
Non compiled:  tensor([0., 0., 1., 0., 0.], device='cuda:0')
Compiled:  tensor([nan, nan, inf, nan, nan], device='cuda:0')
----------------------------------------
```

### Versions

PyTorch version: 2.9.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 12 (bookworm) (x86_64)
GCC version: (Debian 12.2.0-14+deb12u1) 12.2.0
Clang version: 14.0.6
CMake version: version 3.25.1
Libc version: glibc-2.36

Python version: 3.12.0 (main, Oct  3 2023, 01:27:23) [Clang 17.0.1 ] (64-bit runtime)
Python platform: Linux-6.1.0-40-amd64-x86_64-with-glibc2.36
Is CUDA available: True
CUDA runtime version: 12.6.85
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 560.35.05
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           39 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  20
On-line CPU(s) list:                     0-19
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) i9-10900 CPU @ 2.80GHz
CPU family:                              6
Model:                                   165
Thread(s) per core:                      2
Core(s) per socket:                      10
Socket(s):                               1
Stepping:                                5
CPU(s) scaling MHz:                      83%
CPU max MHz:                             5200.0000
CPU min MHz:                             800.0000
BogoMIPS:                                5599.85
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp pku ospke md_clear flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               320 KiB (10 instances)
L1i cache:                               320 KiB (10 instances)
L2 cache:                                2.5 MiB (10 instances)
L3 cache:                                20 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-19
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             KVM: Mitigation: VMX disabled
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Mitigation; Microcode
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] Could not collect
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] torch                     1.12.1+cu116             pypi_0    pypi
[conda] torch-cluster             1.6.0+pt112cu116          pypi_0    pypi
[conda] torch-ema                 0.3                      pypi_0    pypi
[conda] torch-geometric           2.0.3                    pypi_0    pypi
[conda] torch-scatter             2.1.0+pt112cu116          pypi_0    pypi
[conda] torch-sparse              0.6.16+pt112cu116          pypi_0    pypi
[conda] torch-spline-conv         1.2.1+pt112cu116          pypi_0    pypi

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @muchulee8 @amjames @aakhundov @coconutruben

### PR DESCRIPTION
**Pull request by Aminsed**
Fixes #168358

## Summary
When using `index_reduce` or `scatter_reduce` with `reduce='amax'` or `'amin'` and `include_self=False`, the backward pass produces NaN gradients.

## Root Cause
In the backward implementation (`FunctionsManual.cpp`), gradients are distributed among elements that contributed to the max/min result. The code computes `N_to_distribute` (count of elements equal to the result at each position) and divides the gradient by it.

When `include_self=False`:
1. `self_is_result = (self == result)` is 0 for positions where `self` was initialized to ±inf
2. If the result came entirely from `source`, then `N_to_distribute = 0` at that position
3. Division by zero → `inf` → NaN propagation

## Fix
Add `N_to_distribute.masked_fill_(N_to_distribute == 0, 1)` before division in both `scatter_reduce_backward` and `index_reduce_backward` for the `amax`/`amin` case.

This is safe because when `N_to_distribute == 0`:
- Both `self_is_result` and `source_is_result` are 0 at that position
- Multiplying by 0 later gives 0 regardless of the division result
- The fix matches the existing pattern used for the `mean` reduction

## Test Plan
Added `test_scatter_index_reduce_amax_amin_no_nan_grad` which verifies that gradients are finite when `include_self=False` and all values scatter to the same position.

cc @soulitzer @albanD

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
@albanD i missed this notification but did you send to me because you saw distribute in the code? I think this is not actually distributed related.
